rlberry/agents/bandits/index_agents.py

```diff
@@ -0,0 +1,152 @@
+import numpy as np
+from rlberry.agents import AgentWithSimplePolicy
+import dill
+import pickle
+import logging
+from pathlib import Path
+
+logger = logging.getLogger(__name__)
+
+
+class IndexAgent(AgentWithSimplePolicy):
+    """
+    Agent for bandit environment using Index-based policy.
+
+    Parameters
+    -----------
+    env : rlberry bandit environment
+
+    index_function : callable, default = lambda rew, t : np.mean(rew)
+        Compute the index for an arm using the past rewards on this arm and
+        the current time t.
+
+    recursive_index_function : tuple of callable or None
+        Compute recursively a sufficient statistic with the first member of the
+        tuple, the second member is the size of this stat and the last compute the
+        index using the sufficient statistics.
+
+    record_action : bool, default=False
+        If True, record in the writer the number of time an action is played at
+        the end of each simulation.
+
+    phase : bool, default = None
+        If True, compute "phased bandit" where the index is computed only every
+        phase**j. If None, the Bandit is not phased.
+    """
+    name = 'IndexAgent'
+    def __init__(self, env, index_function = lambda rew, t : np.mean(rew),
+                 recursive_index_function=None, record_action = False,
+                 phase = None, **kwargs):
+        AgentWithSimplePolicy.__init__(self, env, **kwargs)
+        self.n_arms = self.env.action_space.n
+        self.index_function = index_function
+        self.recursive_index_function = recursive_index_function
+        self.record_action = record_action
+        self.phase = phase
+    def fit(self, budget=None, **kwargs):
+        n_episodes = budget
+        rewards = np.zeros(n_episodes)
+        actions = np.ones(n_episodes)*np.nan
+        # Initialization : pull all the arms once
+        for a in range(self.n_arms):
```

Comment:
The `fit` method can be called several times, and we'd like to run this loop only the first time, right? So we could keep a total time counter and do something like:

```python
for t in range(n_episodes):
   self.total_time += 1  # initialized to 0 in __init__
   if self.total_time < self.n_arms:
       action = self.global_time
   else:
       ...  # indexes etc

```

Reply:
Yes, good idea. Most of the time, I think in a bandit framework we don't need to continue a fit but I agree it could be useful, I will make the change.